### PR TITLE
MDEV-137: Edit project countries in the Admin Panel

### PR DESCRIPTION
**Pull request by tcaiger**
### Changes:

- Add list of countries to GetProjects endpoint
- Add EditProject api endpoint
- Add countries to project edit form

---

### Screenshots:
Screen Shot 2023-04-27 at 9 24 28 AM

### CODE REVIEW

**Review by acdunham — APPROVED** [review]
Looks fine to me! Nice tests.

File: packages/central-server/src/apiV2/projects/EditProject.js [hunk 1]
```diff
@@ -0,0 +1,84 @@
+/*
+ * Tupaia
+ *  Copyright (c) 2017 - 2021 Beyond Essential Systems Pty Ltd
+ */
+
+import { EditHandler } from '../EditHandler';
+import { assertAnyPermissions, assertBESAdminAccess } from '../../permissions';
+
+export class EditProject extends EditHandler {
```
Comment (EMcQ-BES) [inline packages/central-server/src/apiV2/projects/EditProject.js:9]: Also just acknowledging that @acdunham wrote an `EditProject.js` file which is part of the landing pages epic atm, so this'll cause a pretty major conflict when we have to merge those

File: packages/central-server/src/apiV2/GETProjects.js [hunk 2]
```diff
@@ -39,14 +39,62 @@ export class GETProjects extends GETHandler {
...
 
-    return super.findSingleRecord(projectId, options);
+    const countryNameKey = options.columns.find(x => Object.keys(x)[0] === this.countriesFieldKey);
+    if (!countryNameKey) {
+      return super.findSingleRecord(projectId, options);
+    }
+
+    const columns = options.columns.filter(x => Object.keys(x)[0] !== this.countriesFieldKey);
+    const dbOptions = { ...options, columns: [...columns, { id: 'project.id' }] };
+    const countriesById = await this.getProjectCountriesByProjectId();
```
Comment (EMcQ-BES) [inline packages/central-server/src/apiV2/GETProjects.js:58]: Getting all countries for all projects when only requesting a single record seems excessive? Maybe we can split out a singular version of the `getProjectCountries` function

File: packages/central-server/src/apiV2/projects/EditProject.js [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+
+      if (Object.keys(updatedFields).length > 0) {
+        await this.models.project.updateById(this.recordId, updatedFields);
+      }
+
+      return { id: this.recordId };
+    });
+  }
+
+  async createProjectEntityRelations(models, projectId, countries) {
```
Comment (EMcQ-BES) [inline packages/central-server/src/apiV2/projects/EditProject.js:32]: Again, might be worth breaking this out into the `modelClasses`

```
this.models.project.createProjectEntityRelations(this.recordId, countries);
```

**Review by EMcQ-BES — CHANGES_REQUESTED** [review]
At some point I wanna update our database model to handle through tables better so we don't have to write this much custom code to use them.

The logic itself looks pretty solid, but there's a few spots we might be able to clean-up the code structure

File: packages/central-server/src/tests/apiV2/projects/EditProject.test.js [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+    id: PROJECT_ID,
+    code: 'test_project',
+    description: 'This is a test description',
+    sort_order: 3,
+    image_url: 'www.image.com',
+    logo_url: 'www.image.com',
+    dashboard_group_name: 'test_dashboard',
+  };
+
+  before(async () => {
```
Comment (EMcQ-BES) [inline packages/central-server/src/tests/apiV2/projects/EditProject.test.js:60]: Codeship complaining about this `before` 🤔

File: packages/central-server/src/apiV2/projects/EditProject.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (tcaiger) [reply packages/central-server/src/apiV2/projects/EditProject.js:9]: Inheritance for the win

**Pull request closed without merging**